operator-framework/src/main/java/io/javaoperatorsdk/operator/processing/EventDispatcher.java

```diff
@@ -35,51 +36,84 @@ public EventDispatcher(ResourceController controller,
         this.generationAware = generationAware;
     }
 
-    public void handleEvent(CustomResourceEvent event) {
-        Watcher.Action action = event.getAction();
-        CustomResource resource = event.getResource();
-        log.info("Handling {} event for resource {}", action, resource.getMetadata());
-        if (Watcher.Action.ERROR == action) {
-            log.error("Received error for resource: {}", resource.getMetadata().getName());
-            return;
+    public PostExecutionControl handleEvent(ExecutionScope event) {
+        try {
+            return handDispatch(event);
+        } catch (RuntimeException e) {
+            log.error("Error during event processing {} failed.", event, e);
+            return PostExecutionControl.defaultDispatch();
         }
-        if (markedForDeletion(resource) && !ControllerUtils.hasGivenFinalizer(resource, resourceFinalizer)) {
-            log.debug("Skipping event dispatching since its marked for deletion but does not have finalizer: {}", event);
-            return;
+    }
+
+    private PostExecutionControl handDispatch(ExecutionScope executionScope) {
+        CustomResource resource = executionScope.getCustomResource();
+        log.debug("Handling events: {} for resource {}", executionScope.getEvents(), resource.getMetadata());
+
+        if (ProcessingUtils.containsCustomResourceDeletedEvent(executionScope.getEvents())) {
```

Comment:
Let's say we have a few update events and then a delete event. 
Isn't it better to split the events into the lists with similar action and dispatch them in the coming order? 

One imaginary case could be:
1. update resource to restore a backup
2. update resource to do some cleanup
3. update resource to do get back and put into an object storage
4. Delete the resource.

If they all fall into the same execution scope, if I understood it correctly the resource will be deleted and the first 3 events will be ignored.

Reply:
@csviri pinging because not sure if you saw this comment.